prototypes/ScheduledMerges.hs

```diff
@@ -0,0 +1,551 @@
+               MergingRun _ _ ref -> readSTRef ref
+
+      assert (case mr of
+                SingleRun{} -> True
+                MergingRun mp ml _ -> mergePolicyForLevel ln ls == mp
+                                   && mergeLastForLevel ls == ml)
+        assert (length rs <= 3) $
+        assert (expectedRunLengths ln rs ls) $
+        assert (expectedMergingRunLengths ln mr mrs ls) $
+        return ()
+
+      go (ln+1) ls
+
+    -- All runs within a level "proper" (as opposed to the incoming runs
+    -- being merged) should be of the correct size for the level.
+    expectedRunLengths :: Int -> [Run] -> [Level s] -> Bool
+    expectedRunLengths ln rs ls =
+      case mergePolicyForLevel ln ls of
+        MergePolicyLevelling -> null rs
+        MergePolicyTiering   -> all (\r -> tieringRunSizeToLevel r == ln) rs
+
+    -- Incoming runs being merged also need to be of the right size, but the
+    -- conditions are more complicated.
+    expectedMergingRunLengths :: Int -> MergingRun s -> MergingRunState
+                              -> [Level s] -> Bool
+    expectedMergingRunLengths ln mr mrs ls =
+      case mergePolicyForLevel ln ls of
+        MergePolicyLevelling ->
+          case (mr, mrs) of
+            -- A single incoming run (which thus didn't need merging) must be
+            -- of the expected size already
+            (SingleRun r, CompletedMerge{}) ->
+              assert (levellingRunSizeToLevel r == ln) True
+
+            -- A completed merge for levelling can be of almost any size at all!
+            -- It can be smaller, due to deletions in the last level. But it
+            -- can't be bigger than would fit into the next level.
+            (_, CompletedMerge r) ->
+              assert (levellingRunSizeToLevel r <= ln+1) True
+
+            -- An ongoing merge for levelling should have 4 incoming runs of
+            -- the right size for the level below, and 1 run from this level,
+            -- but the run from this level can be of almost any size for the
+            -- same reasons as above. Although if this is the first merge for
+            -- a new level, it'll have only 4 runs.
+            (_, OngoingMerge _ rs _) ->
+                assert (length rs == 4 || length rs == 5) True
+             && assert (all (\r -> tieringRunSizeToLevel r == ln-1) (init rs)) True
+             && assert (levellingRunSizeToLevel (last rs) <= ln+1) True
+
+        MergePolicyTiering ->
+          case (mr, mrs, mergeLastForLevel ls) of
+            -- A single incoming run (which thus didn't need merging) must be
+            -- of the expected size already
+            (SingleRun r, CompletedMerge{}, _) ->
+              tieringRunSizeToLevel r == ln
+
+            -- A completed last level run can be of almost any smaller size due
+            -- to deletions, but it can't be bigger than the next level down.
+            (_, CompletedMerge r, MergeLastLevel) ->
+              tieringRunSizeToLevel r <= ln+1
```

Comment:
I think you're right actually, for the current (lazy levelling) merge policy it can't happen. I probably wrote this bit before introducing the `SingleRun` case.

Reply:
Flushing under-full write buffers would also hit the previous case (which would need relaxing to allow under-full runs).